### ISSUE TITLE
Async function decorator loses type details if defined as `__call__`

### ISSUE THREAD

**Issue by jakewins** [microsoft/pyright#9677]
**Describe the bug**

The following stateful decorator type fails to pass type info through, seemingly:

```
class StatefulDecorator:
    def __init__(self, some_state):
        self.some_state = some_state

    def __call__(self, fn):
        @functools.wraps(fn)
        async def wrapper(*args, **kwargs):
                return await fn(*args, **kwargs)

        return wrapper
```

If I change it to this and use `this_works` as the decorator call, type checking passes:

```
class StatefulDecorator:
    def __init__(self, some_state):
        self.some_state = some_state

    def this_works(self, fn):
        @functools.wraps(fn)
        async def wrapper(*args, **kwargs):
                return await fn(*args, **kwargs)

        return wrapper
```

**Code or Screenshots**

Full repro:

```
import functools
import typing as t

class StatefulDecorator:
    def __init__(self, some_state):
        self.some_state = some_state

    def __call__(self, fn):
        @functools.wraps(fn)
        async def wrapper(*args, **kwargs):
                return await fn(*args, **kwargs)

        return wrapper

    def this_works(self, fn):
        @functools.wraps(fn)
        async def wrapper(*args, **kwargs):
            return await fn(*args, **kwargs)

        return wrapper

stateful_decorator = StatefulDecorator("some state")


class MyInterface(t.Protocol):
    async def foo(self, x: str) -> str:
        ...



class MyImpl(MyInterface):
    @stateful_decorator
    async def foo(self, x: str) -> str:
        return "bar"


class MyImplWorking(MyInterface):
    @stateful_decorator.this_works
    async def foo(self, x: str) -> str:
        return "bar"
```

```
$ poetry run pyright
repro.py:38:15 - error: "foo" overrides method of same name in class "MyInterface" with incompatible type "_Wrapped[..., Unknown, ..., Coroutine[Any, Any, Unknown]]"
```

**VS Code extension or command-line**

Ran on command line, version `1.1.391`

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

You have failed to provide type annotations for your decorator. In particular, the input parameters have the type `Any`, so pyright is unable to infer the return type. When you apply this decorator to your method, the type information for the method is lost.

You can make your decorator method and/or class generic and use a `ParamSpec` to correctly annotate it and preserve the signature of the decorated method. For details, refer to [this portion of the Python typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#paramspec).

For other general questions about how to use to the Python type system, the [Python typing forum](https://discuss.python.org/c/typing/32) is a great place to get answers.

**Comment by jakewins** [comment]
Haha I'm assuming I'm in the wrong here because I just realized you're literally the author of the typeshed entries for functools.wraps.. https://github.com/python/typeshed/pull/6670/files#diff-02c51ed45b741b6765c365fb5d6578c586510bbbd57b10a1dba10aeba37ae157R85

I remain confused though because:

- The way I read the type signature you wrote for functools.wraps is that is *does* provide that generic type signature and the decorator function itself shouldn't need it?
- I still don't understand why this works for `this_works` but not for `__call__`

**Comment by erictraut** [comment]
The fact that `MyImplWorking.foo` is working is due to pyright's advanced inference capabilities. No other Python type checker would be able to infer the type in this case. I implemented a technique called [call-site return type inference](https://microsoft.github.io/pyright/#/type-inference?id=call-site-return-type-inference), which is able to handle some cases where type annotations are omitted from input parameters, but this has its limits.

If you want your code to work with type checking in general, you should provide type annotations for your input parameters at a minimum. And in cases where your class or method is polymorphic, you should use generics or overloads to describe their behaviors. In the case of `StatefulDecorator.__call__`, you should provide a type annotation for the `fn` parameter. Since this function is generic, you should use a ParamSpec and TypeVar to describe the function.

Here's how to properly annotate your code so it type checks without errors and retains type details.

```python
import functools
from typing import Any, Callable, Coroutine, Protocol

class StatefulDecorator:
    def __init__(self, some_state):
        self.some_state = some_state

    def __call__[**P, R](self, fn: Callable[P, Coroutine[Any, Any, R]]) -> Callable[P, Coroutine[Any, Any, R]]:
        @functools.wraps(fn)
        async def wrapper(*args: P.args, **kwargs: P.kwargs):
            return await fn(*args, **kwargs)

        return wrapper

stateful_decorator = StatefulDecorator("some state")

class MyInterface(Protocol):
    async def foo(self, x: str) -> str: ...

class MyImpl(MyInterface):
    @stateful_decorator
    async def foo(self, x: str) -> str:
        return "bar"
```
